device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,166 @@
+                        break;
+                    }
+                    catch (Exception ex) {
+                        if (ex.getCause() instanceof TransportException && ((TransportException) ex.getCause()).isRetryable()) {
+                            log.warn("Transport exception thrown while opening DeviceClient instance " + getClientId() + ", retrying: ", ex);
+                        } else {
+                            log.error("Non-retryable exception thrown while opening DeviceClient instance " + getClientId() + ": ", ex);
+                            connectionStatus = ConnectionStatus.DISCONNECTED;
+                            throw ex;
+                        }
+                    }
+                }
+            }
+
+            try {
+                log.debug("Sleeping for " + SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS + " secs before attempting another open()");
+                Thread.sleep(SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS * 1000);
+            }
+            catch (InterruptedException ex) {
+                throw new RuntimeException("Interrupted while waiting between attempting to open the client: ", ex);
+            }
+        }
+    }
+
+    public void closeNow() {
+        synchronized (lock) {
+            try {
+                log.debug("Closing the client instance " + getClientId() + " ...");
+                closeClient();
+            }
+            catch (IOException e) {
+                log.error("Exception thrown while closing client instance " + getClientId() + " : ", e);
+            } finally {
+                connectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+    }
+
+    public void open() throws IOException {
+        synchronized (lock) {
+            if(connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+        Pair<IotHubConnectionStatusChangeCallback, Object> suppliedCallbackPair = this.suppliedConnectionStatusChangeCallback;
+        if (throwable == null)
+        {
+            log.info("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason);
+        }
+        else
+        {
+            log.info("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason + ", " + throwable.getMessage());
+            throwable.printStackTrace();
+        }
+
+        if (shouldDeviceReconnect(status, statusChangeReason, throwable)) {
+            if (suppliedCallbackPair != null) {
+                suppliedCallbackPair.getKey().execute(DISCONNECTED_RETRYING, NO_NETWORK, throwable, suppliedCallbackPair.getValue());
```

Comment:
What's the difference between `getClientId()` and `suppliedCallbackPair.getValue()`? Are they both the device Id?

Reply:
in this case, yes, they are both deviceId/multiplexing client Id

Reply:
Seems could be a source of confusion for a sample reader when we use both in the same function.

Reply:
The suppliedCallbackPair is removed. We no longer have this issue.